### PR TITLE
util: add file locking for PCIe access.

### PR DESCRIPTION
**Pull request by ericonr**
This way, only one application using the library can access a given device at a time.

We use a file descriptor pointing to BAR0 (via the resource0 file) because it's the first one to be opened (i.e. we fail early), and because it makes the most sense semantically: the issue with sharing a device between processes is that they will need to access different page indexes, which are controlled by BAR0.

We chose to use Open File Description (OFD) locking because it has more predictable behavior [1] and is POSIX, unlike the flock(2) API.

No locking is added for the serial mode because it's not used in production, and it's believed that the kernel will already report the device as busy when another process has already opened it.

[1] https://lwn.net/Articles/586904/

### CODE REVIEW

**Comment by ericonr** [comment]
@henriquesimoes 
@gustavosr8

File: util/pcie-defs.h [hunk 1]
```diff
@@ -31,6 +31,9 @@ struct pcie_bars {
     pthread_mutex_t locks[NUM_LOCKS];
 
     FILE *fserport;
+
+    /* fd holding OFD-owned lock for BAR resource files */
+    int fd_lock;
```
Comment (ericonr) [inline util/pcie-defs.h:36]: It seems that mmaping the file and keeping the mapping already holds a reference to it, without having to keep the file descriptor itself open. The previous version closed the fd but another process still couldn't lock it.

I don't know if this is a (un)documented linux extension or something standardized that would be okay to rely on. Pending finding out about that before merging this, possibly in a simplified version, without `fd_lock`.

File: util/pcie-open.cc [hunk 2]
```diff
@@ -28,6 +28,27 @@ void configure_mutexes(struct pcie_bars &bars)
...
+{
+    struct flock arg = {
+        /* exclusive lock */
+        .l_type = F_WRLCK,
+        /* lock the whole file */
+        .l_whence = SEEK_SET,
+        .l_start = 0,
+        .l_len = 0,
+        /* application is resposible for setting to 0 */
+        .l_pid = 0,
```
Comment (henriquesimoes) [inline util/pcie-open.cc:41]: `s/resposible/responsible`... I'd consider adding this is specific for OFD locking:

> application is responsible for setting to 0 for OFD locks

**Comment by ericonr** [comment]
I have been informed that the POSIX text doesn't specify if the reference is to the file itself or to the OFD. So for now I might hold off on merging this (or find some reference for the Linux behavior instead)

**Comment by ericonr** [comment]
https://unix.stackexchange.com/questions/784020/is-mmap-holding-a-reference-to-the-ofd-specified-by-posix-or-a-linux-extension

**Pull request merged**